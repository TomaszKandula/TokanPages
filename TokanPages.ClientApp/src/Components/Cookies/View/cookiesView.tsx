import * as React from "react";
import { Backdrop } from "@material-ui/core";
import { ButtonsDto, OptionsDto } from "../../../Api/Models";
import { ViewProperties } from "../../../Shared/Abstractions";
import { GetDateTime } from "../../../Shared/Services/Formatters";
import { ReactChangeEvent } from "../../../Shared/types";

interface Properties extends ViewProperties {
    isClose: boolean;
    hasSnapshotMode: boolean;
    hasCookieConsent: boolean;
    caption: string;
    text: string;
    detail: string;
    loading: string[];
    options: OptionsDto;
    buttons: ButtonsDto;
    canShowOptions: boolean;
    onAcceptButtonEvent: () => void;
    onManageButtonEvent: () => void;
    onCloseButtonEvent: () => void;
    onStatisticsCheckboxEvent: (event: ReactChangeEvent) => void;
    onMarketingCheckboxEvent: (event: ReactChangeEvent) => void;
    onPersonalizationCheckboxEvent: (event: ReactChangeEvent) => void;
    hasStatistics: boolean;
    hasMarketing: boolean;
    hasPersonalization: boolean;
}

const CookieWindowOptions = (props: Properties): React.ReactElement => {
    return (
        <div className="cookie-window-options-list">
            <label className="cookie-window-checkbox pointer-not-allowed">
                <input 
                    type="checkbox" 
                    className="cookie-window-checkbox-input" 
                    disabled={true} 
                    checked={true}
                />
                <span className="cookie-window-checkbox-visual-input"></span>
                <span className="cookie-window-checkbox-label">
                    {props.options?.necessaryLabel}
                </span>
            </label>
            <label className="cookie-window-checkbox">
                <input 
                    type="checkbox" 
                    className="cookie-window-checkbox-input"
                    onChange={props.onStatisticsCheckboxEvent}
                    checked={props.hasStatistics}
                />
                <span className="cookie-window-checkbox-visual-input"></span>
                <span className="cookie-window-checkbox-label">
                    {props.options?.statisticsLabel}
                </span>
            </label>
            <label className="cookie-window-checkbox">
                <input 
                    type="checkbox" 
                    className="cookie-window-checkbox-input"
                    onChange={props.onMarketingCheckboxEvent}
                    checked={props.hasMarketing}
                />
                <span className="cookie-window-checkbox-visual-input"></span>
                <span className="cookie-window-checkbox-label">
                    {props.options?.marketingLabel}
                </span>
            </label>
            <label className="cookie-window-checkbox">
                <input 
                    type="checkbox" 
                    className="cookie-window-checkbox-input"
                    onChange={props.onPersonalizationCheckboxEvent}
                    checked={props.hasPersonalization}
                />
                <span className="cookie-window-checkbox-visual-input"></span>
                <span className="cookie-window-checkbox-label">
                    {props.options?.personalizationLabel}
                </span>
            </label>
        </div>
    );
}

const CookieWindowActions = (props: Properties): React.ReactElement => {
    return (
        <div className="cookie-window-actions">
            {props.buttons?.acceptButton.enabled 
            ? <button 
                className="cookie-window-button cookie-window-button-accent cookie-window-action" 
                onClick={props.onAcceptButtonEvent}
            >
                {props.buttons?.acceptButton.label}
            </button>
            : <></>}
            {props.buttons?.manageButton.enabled 
            ? <button 
                className="cookie-window-button cookie-window-action"
                onClick={props.onManageButtonEvent}
            >
                {props.buttons?.manageButton.label}
            </button>
            : <></>}
            {props.buttons?.closeButton.enabled 
            ? <button 
                className="cookie-window-button"
                onClick={props.onCloseButtonEvent}
            >
                {props.buttons?.closeButton.label}
            </button>
            : null}
        </div>
    );
}

const CookieWindowPrompt = (props: Properties): React.ReactElement => {
    return (
        <div className="cookie-window">
            <div className="cookie-window-caption">
                {props.caption}
            </div>
            <div className="cookie-window-box">
                <div className="cookie-window-section">
                    <p className="cookie-window-section-text">
                        {props.text}
                    </p>
                </div>
                <div className="cookie-window-section">
                    <p className="cookie-window-section-detail">
                        {props.detail}
                    </p>
                    {props.options?.enabled && props.canShowOptions ? <CookieWindowOptions {...props} /> : null}
                    <CookieWindowActions {...props} />
                </div>
            </div>
        </div>
    );
}

const CookieWindowLoading = (props: Properties): React.ReactElement => {
    const maxLength = props.loading?.length-1;
    const dateTime = new Date().toString();
    const formattedDateTime = GetDateTime({ value: dateTime, hasTimeVisible: true });
    return (
        <div className="cookie-window">
            <div className="cookie-window-caption">
                {props.caption}
            </div>
            <div className="cookie-window-box">
                <div className="cookie-window-section cookie-window-section-fixed-height cookie-window-section-left">
                    {props.loading?.map((value: string, index: number) => (
                    <p className="cookie-window-section-text cookie-window-section-left" key={index} >
                        {value.replace("{DT}", formattedDateTime)}
                        {maxLength === index ? <span className="cookie-window-caret"></span> : null}
                    </p>
                    ))}
                </div>
            </div>
        </div>
    );
}

const CookieWindowContainer = (props: Properties): React.ReactElement => { 
    const style = props.isClose ? "cookie-window-close" : "cookie-window-open";
    const transition = props.isLoading ? undefined : 0;
    return (
        <div className={style}>
            <Backdrop className="backdrop" open={true} transitionDuration={transition}>
                {props.isLoading ? <CookieWindowLoading {...props} /> : <CookieWindowPrompt {...props} />}
            </Backdrop>
        </div>
    )
};

export const CookiesView = (props: Properties): React.ReactElement => {
    if (props.hasSnapshotMode) {
        return <div className="cookie-window-open"></div>;
    }

    if (props.hasCookieConsent) {
        return <></>;
    }

    if (props.loading?.length === 0) {
        return <></>;
    }

    return <CookieWindowContainer {...props} />;
}
