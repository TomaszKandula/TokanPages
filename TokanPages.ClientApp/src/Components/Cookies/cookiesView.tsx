import * as React from "react";
import Container from "@material-ui/core/Container";
import Box from "@material-ui/core/Box";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";
import Card from "@material-ui/core/Card";
import CardContent from "@material-ui/core/CardContent";
import CardActions from "@material-ui/core/CardActions";
import cookiesStyle from "./cookiesStyle";
import { Skeleton } from "@material-ui/lab";

interface IBinding
{
    bind: IProperties;
}

interface IProperties
{
    isLoading: boolean;
    modalClose: boolean;
    shouldShow: boolean;
    caption: string;
    text: string;
    onClickEvent: any;
    buttonText: string;
}

const CookiesView = (props: IBinding): JSX.Element => 
{
    const classes = cookiesStyle();

    const ActiveButton = (): JSX.Element => 
    {
        return(
            <Button onClick={props.bind?.onClickEvent} className={classes.button}>
                {props.bind?.buttonText}
            </Button>
        );
    }

    const renderConsent = (): JSX.Element => 
    {
        return (		
            <Box position="fixed" width="100%" bottom={0} p={2} zIndex="modal" className={props.bind?.modalClose ? classes.close : classes.open}>
                <Container maxWidth="md">
                    <Card elevation={8} className={classes.container}>
                        <CardContent>
                            <Typography gutterBottom={true} className={classes.caption}>
                                {props.bind?.isLoading ? <Skeleton variant="text" /> : props.bind?.caption}
                            </Typography>
                            <Typography className={classes.text}>
                                {props.bind?.isLoading ? <Skeleton variant="text" /> : props.bind?.text}
                            </Typography>
                        </CardContent>
                        <CardActions>
                            {props.bind?.isLoading ? <Skeleton variant="rect" /> : <ActiveButton />}
                        </CardActions>
                    </Card>
                </Container>
            </Box>
        );
    }

    return (<>{props.bind?.shouldShow ? renderConsent() : null}</>);
}

export default CookiesView;
