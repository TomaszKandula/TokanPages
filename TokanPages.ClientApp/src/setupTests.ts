// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import enableHooks from "jest-react-hooks-shallow";

enableHooks(jest);

jest.mock("react-redux", () => (
{
    useSelector: jest.fn(),
    useDispatch: jest.fn()
}));

jest.mock("connected-react-router", () => (
{
    connectRouter: jest.fn(),
    routerMiddleware: jest.fn()
}));

Enzyme.configure({ adapter: new Adapter(), disableLifecycleMethods: false });

// environmental variables used during all tests
process.env.REACT_APP_API_VER="1";
process.env.REACT_APP_BACKEND="http://localhost:5000";
