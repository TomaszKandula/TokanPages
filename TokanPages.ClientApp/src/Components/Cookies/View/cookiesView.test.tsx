import "../../../setupTests";
import React from "react";
import { render } from "@testing-library/react";
import { CookiesView } from "./cookiesView";

describe("test component: cookiesView", () => {
    it("should render correctly '<CookiesView />' when content is loaded.", () => {
        const html = render(
            <CookiesView
                isLoading={false}
                modalClose={false}
                shouldShow={false}
                caption="Cookie Policy"
                text="We use cookies to personalise content..."
                detail="With your consent, we may also..."
                onClickEvent={jest.fn()}
                options={{
                    enabled: true,
                    necessaryLabel: "Necessary",
                    statisticsLabel: "Statistics",
                    marketingLabel: "Marketing",
                    personalizationLabel: "Personalization"            
                }}
                buttons={{
                    acceptButton: {
                        label: "Accept All",
                        enabled: true
                    },
                    manageButton: {
                        label: "Manage Settings",
                        enabled: true
                    },
                    closeButton: {
                        label: "Close",
                        enabled: true
                    }
                }}
            />
        );

        expect(html).toMatchSnapshot();
    });
});
