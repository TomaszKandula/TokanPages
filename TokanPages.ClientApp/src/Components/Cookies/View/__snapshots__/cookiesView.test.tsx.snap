// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test component: cookiesView should render correctly '<CookiesView />' when content is loaded. 1`] = `<Fragment />`;
